### ISSUE TITLE
Train with FP

### ISSUE THREAD

**Issue by iris-ywang** [iris-ywang/chemvae_model#4]
Train the SAE and PAE using Fingerprints instead of SMILES.
- Attempt 1: Train with the initial experiment setup using Morgan fingerprint. 
- Required modification: Change the first and last layer to the required size (1024,). Decoder output is in the reversed order.

- Attempt 2: Train the SAE/PAE on ZINC with fingerprints (which could
be generated from SMILES), then QSPR-test it on ZINC held-out test data AND QSAR-test it on other datasets (e.g ChEMBL).

**Comment by iris-ywang** [comment]
Need to check with Abbi how the CHEMBL data was generated. Using the method suggested by chat GPT, I can't regenerate the fingerprints in CHEMBL datasets. (Plus the dense 1024 bits even have bits=2 : (

**Comment by iris-ywang** [comment]
Use CHEMBL dataset first.

**Comment by iris-ywang** [comment]
**MFP 1024 Input Network shape change:** 

(1024, 1) -> Conv(1016, 9) -> BatchNorm(1016, 9) -> Conv(1008, 9) -> BatchNorm(1008, 9) -> Conv(998, 9) -> BatchNorm(998, 9) -> Flatten(9980) -> Dense(196)

**SMILES 120x35 Input Network shape change:** 

(120, 35) -> Conv(112, 9) -> BatchNorm(112, 9) -> Conv(104, 9) -> BatchNorm(104, 9) -> Conv(94, 9) -> BatchNorm(94, 9) -> Flatten(940) -> Dense(196)

Note, the original plan ((120, 35) -> 196) reduces the Flattened layer of size 940 to 196 (**~5x**). But MFP version reduces the size from 9980 to 196 (**~51x**).

**Comment by iris-ywang** [comment]
Running on CSD3. Runnable BUT:

- For FP VAE, 10% of a batch takes 2 hours
- For SMILES VAE, one batch takes 6 hours
- Note on CEB server, one SMILES VAE batch takes 15 min.

Suspecting that the script is not using GPU to run. It was running on CPU.

**Comment by iris-ywang** [comment]
To compare with CEB server, re-run CHEMBL VAE.

Necessary change: 
a. bottom of train_vae.py, use the zinc/exp.json line
b. disable vectorise_chembl_data, enable vertorise_data
c. In models/zinc/exp.json, add `/home/yw453/chemvae_model/` instead of `../`

Attempt remedy:
1. In `run_wilke3_smiles_test.sh`, add:
a. module load cuda/11.8
b. nvidia-smi 

2. In `train_vae.py`, add:
a. print GPU info during the task and force it to use GPU
```
# Set up GPU info
gpus = tf.config.list_physical_devices('GPU')
if gpus:
    logging.info(f"GPUs available: {gpus}")
else:
    logging.info("No GPU available!")


if gpus:
    try:
        # Set memory growth for each GPU (prevents TensorFlow from using all the GPU memory at once)
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)

        # Make all GPUs visible to TensorFlow
        tf.config.set_visible_devices(gpus, 'GPU')

        print(f"Using GPUs: {gpus}")
    except RuntimeError as e:
        print(f"Error setting GPUs: {e}")
```

b. add gpu_callbacks to the callbacks list (not sure if it's going to work)
```
def print_gpu_usage():
    gpu_usage = subprocess.run(['nvidia-smi', '--format=csv', '--query-gpu=utilization.gpu,memory.used'], stdout=subprocess.PIPE)
    print(gpu_usage.stdout.decode('utf-8'))

# Define a custom callback to print GPU usage
class GPUUsageCallback(Callback):
    def on_train_batch_end(self, batch, logs=None):
        print(f"Batch {batch} GPU usage:")
        print_gpu_usage()
```
and 
```
    gpu_usage_cb = GPUUsageCallback()
    callbacks = [vae_anneal_callback, csv_clb, gpu_usage_cb]
```


Test run submitted to: 62242440

**Comment by iris-ywang** [comment]
Emailed CSD3 staff, and they said Ampere doesn't have cuda=11.8

- FIXED on 22/10/2024! Inspiration! I just need to use lower grade of tensorflow-cuda-cudnn!

Worked on HPC with $module load python/3.11 cuda/11.2 cudnn/8.1_cuda-11.2 and tensorflow==2.11.0!

On HPC ZINC SA-VAE takes 5 min per batch. 

On CEB ZINC SA-VAE takes 35 min per batch.

So there is a speed benefit for sure.

**Comment by iris-ywang** [comment]
Actually the comment above is not true - on both machines it's only taking one GPU to run, so they are on the same speed. The discrepancy I wrote is just a miss-read from logging time...

**Comment by iris-ywang** [comment]
Have been running FP-paired on CHEMBL204 on CEB Server over the weekend. The NaN weights problem occurred again after 6 batches (next time run from batch = 6, reload=True). 

To-do:
1. Reducing the value of clipnorm again like last time. 
2. Try https://www.tensorflow.org/guide/keras/distributed_training on CEB server to use multiple GPUs

**Comment by iris-ywang** [comment]
Timing wise: if using one GPU, 7 hours per batch for FP-Paired (as in 7hr / 200 samples)
